src/index.ts

```diff
@@ -216,8 +247,18 @@ export class Ok<T> implements BaseResult<T, never> {
     safeUnwrap(): T {
         return this.val;
     }
+
+    flatMap<T2, E2>(mapper: (val: T) => Err<E2>): Err<E2>;
+    flatMap<T2>(mapper: (val: T) => Ok<T2>): Ok<FlattenResults<T2>>;
+    flatMap<T2, E2>(mapper: (val: T) => any): any {
```

Comment:
How often, if ever, will we need to recursively flatten here? Recursively flattening really complicates the typing, and I can't imagine it being very common.

For instance, if we do support recursively flattening, we will need to add a test to verify the return type of this:
```typescript
let x: Result<T, E>;
let y: Result<Result<T2, E2>, E3>;
x.flatMap(() => y); // This return type
```
Is equal to this:
```typescript
Result<T2, E | E2 | E3>
```

My vote would be to simplify by declaring the function like this:
```typescript
	flatMap<E2>(mapper: (val: T) => Err<E2>): Err<E2>;
	flatMap<T2>(mapper: (val: T) => Ok<T2>): Ok<T2>;
	flatMap<T2, E2>(mapper: (val: T) => Result<T2, E2>): Result<T2, E2> {
		return mapper(this.val);
	}
```

If you are commonly seeing more than 2 layers of nested `Result`s, I'd be happy to change my mind on this.  Alternatively, we could add a `flatten()` function that would help with that scenario.

Reply:
Hmm, imho that's the main if not only purpose of flatMap - to recursively unwrap the result, otherwise I don't really see any advantage over a simple map.

Or am I l misunderstanding your argument?

Also, the flatten is a nice idea, that could work smtg like this:

`
nestedResult.map(flatten).((deepestNestedVal) => //do stuff).
`

Which can be replaced by a flatmap for sugar.

Reply:
Say that I have this function:
```typescript
function loadThenParse(url: string): Result<Result<ParsedObj, ParseErr>, NetworkErr> {...}
```

And then I call flatmap onto that function:
```typescript
const urlResult: Result<string, UrlErr> = new Ok("myUrl");
const parsed = urlResult.flatMap(url => loadThenParse(url));
```

`parsed` can have two possible types:
```typescript
// This is if flatMap recursively unwraps results
const parsed: Result<ParsedObj, UrlErr | ParseErr | NetworkErr> = ...;

// This is if flatMap only unwraps one layer of results
const parsed: Result<Result<ParsedObj, ParseErr>, UrlErr | NetworkErr> = ...;
```

99% of the time having `flatMap` unwrap one layer of results is all that is needed.  In the strange scenario where you are calling `flatMap` onto nested results, like the one above, my vote is that we only unwrap one layer.  This is for a few reasons:
* It's much simpler to add Typescript typings to, which also makes the api easier to consume
* It gives explicit control of flattening to the user.  They may or may not want to completely flatten.
  * If they want to flatten both layers, they can simply call `flatMap` again.  
  * It'd be even cleaner if we added a `flatten()` function.

I'm willing to do the recursive unwrapping if somebody can show me a common use-case for it, but I don't think the added complexity is worth it for the tiny percent of the time where it could be useful

Reply:
I'll add more examples when I'll have some time a bit later, but just a high level argument that I want to make here is that imho what you're describing isn't flatMap, in which case it will be confusing to consumers. 

I understand the need for simplifying the types around it but am not 100% sure yet that that's a real limitation in terms of complexity.

If we are going along these lines then I suggest we change the name. `deepMap` sounds the most feasible for now for me with the benefit that we could manually type it for say up to 5-10 levels and name it `deepMap2`, `deepMap3`, etc... This would add the benefit of being much **simpler/faster at compile time** and that **the user now controls the levels of unwraping.** 

But on the other hand:
1. I don't remember ever seeing some library implementing a `deepMap`, but I do know about at least a handful of libraries and concepts using `flatMap`. My argument here is that `flatMap` is already part of the ecosystem and has an expected/known behavior across many platforms and concepts, so it would be easier to adopt.
2. Controlling the level of wrapping at the consumer level, sounds like something needed but I in my experience in most of the case I really don't and shouldn't care about how many levels deep my data is – I just want it out so I can play with it. Think about Promises, which implement `flatMap` by default in every `then`. I don't think I ever needed a `then` to give me a Promise Object instead of the unwrapped value.
3. In fact I would argue that not knowing how many levels of wrapping exists in my result before getting the data out of it, up in my high level consumer function, is not only easier but also a characteristic of Cleaner Code, since all I need to know up there is _what_ my input looks like not _how_ to get it!

Let me know your thoughts around all of these and let's keep up the conversation. I think it's great 👍

Reply:
The reason Promises automatically unwrap is that doing anything else would have no purpose.  If you were to unwrap one level of a `Promise<Promise<string>>`, there's nothing you could ever want to do with the `Promise<string>` other than unwrap it.

An example of another system that has `flatMap` is rxjs's `Observable`.
```typescript
let x: Observable<string>;
let y: Observable<Observable<number>>;
let z: Observable<Observable<number>> = x.flatMap(() => y);
```

Another example can be found in the built in [`Array.flatMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap):
```typescript
let x: Array<string>;
let y: Array<Array<number>>;

let z: Array<Array<number>> = x.flatMap(() => y);
```
You can try this in the typescript playground [here](https://www.typescriptlang.org/play?target=99#code/DYUwLgBAHgXBCCAnRBDAngHgM5kQSwDsBzAPggF4IBtAchRoF0BuAKFEjTiVU2-QwIBXALYAjEIhJlKVKgEYANBABMSgMwMlVACxKArEoBsDZizbgIALy7J+fTELESpFaADoAZsBRgAsigAHAAoggEoKMjRQ1gBjAHsCLDjQN2A4oiDLaKA).

The reason `Observable.flatMap` and `Array.flatMap` don't recursively flatten is that there is that the user may only wish to do one layer of flattening.  `Promise` recursively flattens because there is no scenario when you wouldn't want that.

`Result.flatMap` should behave the same way, giving the user control on how much they want to flatten.  Similar to the [`Array.flat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) function, we should provide a `Result.flatten` function.  This gives the user full control, greatly simplifies the types, and follows existing Javascript flatMap patterns. 

Again, 99% of the time the behavior will be the exact same as they won't be calling `Result.flatMap` on deeply nested results.

Reply:
I see. Good points about the Rx and Array `flatmap` behavior. Seeing that made me change my mind, but to play the devil's advocate till the end, do you have any use cases where you'd want to work with the Result instead of the unwrapped value?

You're saying "_if you were to unwrap one level of a Promise<Promise<string>>, there's nothing you could ever want to do with the Promise<string> other than unwrap it._", but I can't really think of a case where I'd need to work with the Result either. In my mind the Result concept/behavior is related more to Promises than Arrays.

One more thing we can consider, very similar to the `deepMap` I tried to describe above, we can take a page from the [Array.flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) and add depth levels to the `flatMap`, or `flatten` method.

Let me know what you think.

Reply:
I am having a very hard time coming up with any use case where you'd have a `Result<Result<T, E>, E2>` in your code at all.  Any place where that could arrive you could simply use flatMap and have a flattened `Result`.  

The only reason I can think of is if you truly have two separate error types that the caller may or may not want to handle separately.

In the end it boils down to this: recursively flattening seems like it would _rarely_ be used, and isn't obviously beneficial in all cases.  The added complexity to the typescript interface to add recursive unwrapping seems unjustified.  

I love the idea of adding depth levels to `flatten` similar to `Array.flat`

Reply:
Hey @vultix, sorry for the late reply. I actually came to agree with you more and more. I've been using the 1level flatMap in my code and you're right – the times when you'd need a deeper level is very rare. 

I'd say let's go with the 1level approach for now, which is simpler to implement and maybe add the nested levels argument later on.

I'll be able to make the changes sometime in the next week or so, but if you are eager to work on this earlier than that please do so. 

P.S. I just stumbled upon the new `andThen()` functionality and from what I quickly saw w/o testing it I can see it pretty much implements the 1 level `flatMap` behavior. Is the difference that `andThen()` would act more like a `finally()` where both `mapErr` and `map` would end up calling it?